querier.go

```diff
@@ -0,0 +1,365 @@
+					Buckets: []float64{0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 4.5, 5, 6, 7, 8, 9, 10},
+				},
+				[]string{"grpc_code", "range"},
+			),
+			singleHistogram: promauto.With(reg).NewHistogramVec(
+				prometheus.HistogramOpts{
+					Name:        "parca_client_query_seconds",
+					Help:        "The seconds it takes to make Query requests against a Parca",
+					ConstLabels: map[string]string{"mode": "single"},
+					Buckets:     prometheus.ExponentialBucketsRange(0.1, 120, 30),
+				},
+				[]string{"grpc_code", "report_type", "range"},
+			),
+			mergeHistogram: promauto.With(reg).NewHistogramVec(
+				prometheus.HistogramOpts{
+					Name:        "parca_client_query_seconds",
+					Help:        "The seconds it takes to make Query requests against a Parca",
+					ConstLabels: map[string]string{"mode": "merge"},
+					Buckets: []float64{
+						0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
+						12, 13, 14, 15, 16, 17, 18, 19, 20,
+					},
+				},
+				[]string{"grpc_code", "report_type", "range"},
+			),
+		},
+		client:          client,
+		series:          make(map[string]timestampRange),
+		queryTimeRanges: []time.Duration{15 * time.Minute, 7 * 24 * time.Hour},
+		reportTypes: []queryv1alpha1.QueryRequest_ReportType{
+			queryv1alpha1.QueryRequest_REPORT_TYPE_PPROF,
+			// nolint:staticcheck
+			queryv1alpha1.QueryRequest_REPORT_TYPE_FLAMEGRAPH_UNSPECIFIED,
+			queryv1alpha1.QueryRequest_REPORT_TYPE_FLAMEGRAPH_TABLE,
+		},
+	}
+}
+
+func (q *Querier) Run(ctx context.Context, interval time.Duration) {
+	ctx, cancel := context.WithCancel(ctx)
+	q.cancel = cancel
+
+	defer close(q.done)
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// Since a lot of these queries are dependent on other queries, run
+			// them one at a time.
+			q.queryLabels(ctx)
+			q.queryValues(ctx)
+			q.queryProfileTypes(ctx)
+			q.queryRange(ctx)
+			q.querySingle(ctx)
+			q.queryMerge(ctx)
+		}
+	}
+}
+
+func (q *Querier) Stop() {
+	q.cancel()
+	<-q.done
+}
+
+func (q *Querier) queryLabels(ctx context.Context) {
+	queryStart := time.Now()
+	resp, err := q.client.Labels(ctx, connect.NewRequest(&queryv1alpha1.LabelsRequest{}))
+	latency := time.Since(queryStart)
+	if err != nil {
+		q.metrics.labelsHistogram.WithLabelValues(connect.CodeOf(err).String()).Observe(latency.Seconds())
+		log.Println("labels: failed to make request", err)
+		return
+	}
+	q.metrics.labelsHistogram.WithLabelValues(grpcCodeOK).Observe(latency.Seconds())
+	q.labels = append(q.labels[:0], resp.Msg.LabelNames...)
+	log.Printf("labels: took %v and got %d results\n", latency, len(resp.Msg.LabelNames))
+}
+
+func (q *Querier) queryValues(ctx context.Context) {
+	if len(q.labels) == 0 {
+		log.Println("values: no labels to query")
+		return
+	}
+	label := q.labels[len(q.labels)/2]
```

Comment:
This will always query values for the same label name if the labels are stable. We should have some randomness added here, such that every run we query slightly different data in the system.

Reply:
Agreed, I'm going to add in an rng. The only problem is that we could report metrics for different queries. In this case it doesn't matter but what about merging cpu vs memory profiles? Not sure if that should belong on the same time series. Maybe we should do a query per value and have those be different labels in time series?

Reply:
Yes, fair point. I think just randomness has served us well up until now, but we can always change that.